### ISSUE TITLE
sqrt prevents evaluation

### ISSUE THREAD

**Issue by georgekinnear** [maths/moodle-qtype_stack#824]
One of my questions broke when STACK was upgraded to the latest version - I'm guessing because of something in the parser, as I've run in to problems before with "unclear substitution".

Here's a minimal example based on my question (code below, and exported question here: [sqrt-substitution.zip](https://github.com/maths/moodle-qtype_stack/files/9232202/sqrt-substitution.zip)). I first want to solve for a particular value of t, saved as `myvar1`. Then I want to evaluate an expression with that value of t.
```
a:rand_range(1,3);
u:matrix([t,-1,a*t]);
v:matrix([-1,a+1,5]);
myvar1:ev(t,solve(u.u=1,t));
myvar2:ev(sqrt(u.v), t=myvar1);
```

On the question tests page, the "Question variable values" field shows:
```
a:1;
u:matrix([t,-1,t]);
v:matrix([-1,2,5]);
w:matrix([1,4,1]);
myvar1:0;
myvar2:sqrt(2)*%i;
```
I see the same value of `myvar2` when I use the "Send to the CAS" button.

However, if I embed {@myvar2@} in the question text, it shows as \(\sqrt{4t-2}\), i.e. without the value of t having been evaluated.

### Workaround
I've got my question working again by avoiding use of `sqrt` inside `ev`:
```
udv:ev(u.v, t=myvar1);
myvar2:sqrt(udv);
```
However this is bit more unwieldy, and maybe won't be possible in all situations.

### Side issues
Aside from my question breaking, I think there are two other problems here:

1. There was no error message. I only spotted this because it led to some question tests failing, since myvar2 was the answer to one of the parts. If something like this is happening in a worked solution, I have no way to know.
2. Debugging was very difficult. The CAS chat and the question tests page both suggested this evaluation was working in the same way it always had. I'm surprised that those are not in sync with what is actually happening in the question.

**Comment by georgekinnear** [comment]
Just to add, I had another broken question with a similar issue. For that one, the use of `rhs` inside `ev` was the issue.

**Comment by aharjula** [comment]
Not helpful, but an update to the general "unclear substitution saga". Back in the last releases of 4.3.x we had a new static analysis-based security system that tried to track down substitutions and detect cases where bad things happened behind those substitutions. It had nothing to do with parsing, but it happened at the same time. The problem was that some expressions were impossible to process due to combinatorial explosions.

4.4 brings a new system that checks values at runtime and has absolutely no issues with anything being unclear. And we no longer have that paranoid/explosive and overly sensitive static analysis system in play, so no "unclear substitutions". Now the problem is more with the new CASText2 system. In the old system, the securely evaluated value was first transferred from the CAS back to PHP and then pushed back into the CAS so that it could be turned into something that gets injected into the CASText. Now, however, the values are evaluated in the same CAS session and are not moved around and serialised (serialisation typically also evaluates things) for transfer, thus leading to a situation where things that have not actually been evaluated all the way are being passed to CASText for rendering, and CASText does everything it can to avoid evaluating them in case someone wants to display unevaluated content (e.g. model solutions and steps).

On the question test page, that listing behaves pretty much the way the old CASText did, i.e. the value has been serialised and rendered in that form.

In general, it seems like we have some issues related to `ev` combined with the substitutions coming from a function. In particular, in your examples `myvar1` line where the substitution is the result of the `solve`. There have been cases where using `at` instead of `ev` has worked, and it does have the same order of arguments, so it can be a quick fix.

As an unrelated comment about that `myvar1`-line I strongly suggest not doing that like that. It is very common for new authors to copy that pattern and not notice that `solve` might return more than one solution (or none at all) if they copy it into a different type of question and not necessarily even in the same order every time. In general, I recommend `myvar1:rhs(first(solve(...)))` as the pattern for extracting the value for usage as it is guaranteed to error out if no value is available or if there happens to be an infinite number of values.

**Comment by aharjula** [comment]
Or basically, `ev` with substitutions coming from functions within the `ev` call seems to be broken. One could even say that those are actually "unclear substitutions". I would not be surprised if things would simply work if you wrote this:

```
...
myvar1:ev(t,solve(u.u=1,t));
...
```

as:


```
...
sol:solve(u.u=1,t);
myvar1:ev(t,sol);
...
```
Basically, evaluate the evil function outside the `ev`.

**Comment by georgekinnear** [comment]
Thanks for the explanation of what's going on.

I tested your suggestion of finding `myvar1` in two steps - but that is not actually the issue. Both your version and my original version are evaluating `myvar1` to a numerical value (I put {@myvar1@} in the question text to check).

The issue is with `myvar2`. I've just tried exploring the boundaries of what works and what doesn't, by adding the following to my question text:

```
<p>{@ev(sqrt(u.v), t=myvar1)@}</p>
<p>{@ev(u.v, t=myvar1)@}</p>
<p>{@ev(sqrt([t,-1,a*t].[-1,a+1,5]), t=myvar1)@}</p>
```
The first line is the original definition of `myvar2` and it shows in the question text with t unevaluated. However, the other two lines show with t being evaluated correctly. So `ev` seems to work when used on something "not too complicated".

Thanks for the tip about `at` - that does indeed work as a drop-in replacement.

Thanks also for the tip about using `rhs(first(solve(...)))`.

**Comment by aharjula** [comment]
Update on this. Apparently, if there is even one function call directly visible in the content that is being evaluated by `ev` then the way that the new security system acts breaks things down. There is no general workaround as one can do so many different things with `ev`, for most uses, the current workarounds are:

 1. `at` if it works, again depends on what exactly is being done.
 2. Using a temporary variable so that the security evaluation can be done outside the `ev`.
 ```
 /* Instead of this */
 var: ev(complexExpression, ...);
 /* Try this, depending on what you are doing, adjusting simp may 
    not be necessary, for most people, it is not necessary */
 simp: false;
 tmp: complexExpression;
 simp: true;
 var: ev(tmp, ...);
 ```
  3. If you are doing things that modify expressions, you might need to use `subst` instead of `ev` it is in any case the better tool for those things.

Still looking into whether we can automatically apply that second solution without the author having to care at all or whether doing so would break too many questions.

Above by "security evaluation" we mean additional calls and conditions we generate that surround the expression and get evaluated before the expression and will error out should they detect something evil.

"not too complicated" is now defined as an expression that does not have function calls that would contain the variables that are being substituted.

**Comment by aharjula** [comment]
Yet another update. Still trying to select the solution that causes the least amount of side effects.

Apparently, adding `eval` to that `ev` call also works for the vast majority of problem cases. It, however has two edge cases where it would not be ideal:
 1. Substitututions that include the target on both sides e.g. `ev(x,x=x+1)`. For these, the use of `subst` is highly recommended. Unfortunately, we cannot easily detect these if the other side is hidden behind something else. But direct ones we can probably detect and warn about.
  2. The use of `noeval` might be the one that we cannot support at all if we go for this fix.
  
So basically, this should also work:
```
myvar2:ev(sqrt(u.v), t=myvar1, eval);
```
